platform/fdo/cog-platform-fdo.c

```diff
@@ -645,6 +649,64 @@ output_handle_scale (void *data,
     g_info ("Got scale factor %i for output %p\n", factor, output);
 }
 
+static bool
+cog_fdo_does_image_match_win_size(struct wpe_fdo_egl_exported_image* image)
+{
+    return image && wpe_fdo_egl_exported_image_get_width(image) == win_data.width &&
+        wpe_fdo_egl_exported_image_get_height(image) == win_data.height;
+}
+
+static void
+cog_fdo_fullscreen_image_ready()
+{
+    xdg_toplevel_set_fullscreen(win_data.xdg_toplevel, NULL);
```

Comment:
This only covers the case in which we are using the `xdg_wm_base` Wayland protocol. The `xdg_toplevel` member will be `NULL` in case we are using the `wl_shell` or `zwp_fullscreen_shell_v1`.

The order in which we prefer them is always this:

1. `xdg_wm_base`
2. `zwp_fullscreen_shell_v1`
3. `wl_shell`

You will need something like this:

```suggestion
    if (wl_data.xdg_shell) {
        xdg_toplevel_set_fullscreen(win_data.xdg_toplevel, NULL);
    } else if (wl_data.fshell) {
        return;  // Always fullscreen, nothing to do (?)
    } else if (wl_data.shell) {
        wl_shell_surface_set_fullscreen (win_data.shell_surface,
                                         WL_SHELL_SURFACE_FULLSCREEN_METHOD_SCALE,
                                         0,
                                         NULL);
    } else {
         g_assert_not_reached();
    }
```

...and similarly to un-fullscreen.

Reply:
(If you ever wondered: yes, having to do `if`-`else` chains in every place where a Wayland “shell” operation is needed gets old and boring soon; I have ideas to improve that, see #336)

Reply:
I added support for wl_shell and zwp_fullscreen_shell_v1 in the latest commit. I had to save the size before fullscreen to be able to get the previous state after exiting fullscreen. It seems that xdg_shell handles that automatically but not the others.

BTW I also noticed that the code uses a space before the `()`of each function call or declaration. I updated the new code to follow the same coding style.

Reply:
We are slowly migrating to _not_ have a space before the function call parentheses, but only for new and modified code. That is why the code style checker failed.

If you have `clang-format` installed locally, you can run the following in your branch to apply the changes suggested by the style checker:

```sh
data/check-style master | git apply -p0
```

Reply:
Thanks for sharing that command!

I updated the code with the fixed styles and the function renames suggested by @carlosgcampos in https://github.com/WebPlatformForEmbedded/libwpe/pull/86